### ISSUE TITLE
Roaming movement in combat-trainer

### ISSUE THREAD

**Issue by brycole** [rpherbig/dr-scripts#1638]
Hi Guys,

I'm planning on adding some roaming movement in combat-trainer but I am having some issues understanding some of the combat-trainer code.

Specifically in combat-trainer, I'm not sure how all the processes pull together. And where would be a good place to inject the movement when no monsters are present.

I've also adapted a bit of the hunting-movement code and not quite understanding how lambda works.

```
def roam()
  roam_last_room = nil

  roam = proc {
    room = Room.current
    # placeholder for room boundary
    #next_room_options = room.wayto.keys - ROOM BOUNDARY ARRAY
    next_room_options = room.wayto.keys
    if next_room_options.length > 1
      next_room_options.delete_if { |option| option == roam_last_room }
    end
    next_room = next_room_options[rand(next_room_options.length)]
    way = room.wayto[next_room]
    if way.class == String
      sleep 0.3
      move(way) if way
    else
      way.call if way.inspect !~ /go root/
    end
    roam_last_room = room.id.to_s
  }

  while true  # roam, check for players
    roam.call
    sleep 0.1
    return if empty_room?
  end
end

**def empty_room?()
  UserVars.friends = @settings.hunting_buddies || []
  lambda do
    return true if (DRRoom.pcs & UserVars.friends).any?
    return false if (DRRoom.pcs - DRRoom.group_members).any?
    # No visible friends in the room and no visible people
    UserVars.friends.each { |friend| Flags.add("room-check-#{friend}", friend) }
    Flags.add('room-check', 'says, ', 'say, ', 'You hear')
    bput('search', 'roundtime')
    data = reget(40).reverse.take_while { |x| x !~ /You search around/ }
    if data.grep(/vague silhouette|You notice \w+, who is|see signs that/).any?
      pause
      waitrt?
      return UserVars.friends.find { |friend| Flags["room-check-#{friend}"] }
    end
    fput("say #{@settings.empty_hunting_room_messages.sample}")
    20.times do |_|
      pause 0.5
      return true if UserVars.friends.find { |friend| Flags["room-check-#{friend}"] }
      return false if Flags['room-check'] || !(DRRoom.pcs - DRRoom.group_members - UserVars.friends).empty?
    end
    true
  end
  true
end

roam
echo "Room empty!"
```

So these pieces actually move but the Lamda section doesn't perform any of the desired functions. 

Thanks!

**Comment by jandersson** [comment]
You should look at `bescort` it has support for what you're trying to do i believe.

**Comment by rcuhljr** [comment]
The core of how the processes are run is here.

```
def start_combat
    @equipment_manager.empty_hands
    loop do
      @combat_processes.each do |process|
        @safety_process.execute(@game_state)
        break if process.execute(@game_state)
      end
      pause 0.1
      if @game_state.done_cleaning_up?
        echo('CombatTrainer::clean_up') if $debug_mode_ct
        @running = false
        stop_script('tendme') if Script.running?('tendme')
        break
      end
      next unless @stop && !@game_state.cleaning_up?
      @game_state.next_clean_up_step
      @game_state.stop_weak_attacks
      @game_state.stop_analyze_combo
    end
  end
```

@combat_processes is the list of processes that make up the combat priority, they're defined in `def make_processes`. Each process gets it's execute method called each loop, unless a process returns true which short circuits the loop back to the top. This is normally used an early process is taking an action that should block lower priority actions.  If I wanted to add roaming I'd probably at a new process early in the pipeline that when combat-trainer receives a 'pause yourself' signal waits until no magic is being cast then begins returning true to block the later processes. Once the pause state is entered hunting buddy can navigate to a new room and send the unblock signal to combat-trainer, which resumes normal functioning.

In this model hunting buddy would watch for some amount of time going by without any mobs in the room, or whatever your criteria is.

**Comment by intangibles** [comment]
Very nice.  Thank you for the explanation. @rcuhljr.   I don't necessarily mind if it had to go through all the processes to check for an empty room. That stuff needs to get trained too!  I was thinking of a good spot to also insert a type of before/after script to do when no monsters are present.   So if roam was off, but no monsters present i might run a scholarship script or read a first aid book.   They sort of tie in together.  (by finding a good place to inject zero mobs.)

What would you think about injecting it near the loot/search phase? That's one spot where you absolutely can predict where it drops from 1 mob to zero mobs.  Maybe that could flip the "pause" switch to fastforward to bypass the other processes?  Or just stick the movement directly after the loot/search phase and bypass modifying the processes?

**Comment by MahtraDR** [comment]
@rpherbig Similar to the other issue, is it worthwhile keeping this 4yo discussion open? It hasn't been enoug of a problem for anyone to comment since. 

PS. I'm going through issues from oldest to newest, tackling what I can quickly, and triaging what I can't solve quickly into a list I can tackle after the initial pass. Is that helpful, or just adding noise to your inbox?

**Comment by rpherbig** [comment]
> Is that helpful, or just adding noise to your inbox?

I don't mind, as long as you find it helpful.

I think newest-first is most likely to solve real problems experienced by our users today. Oldest-first is most likely to find issues that can be closed as stale and clean up the backlog so there is less noise. Either one is useful and I suggest you go with what you find the most interesting.

**Comment by MahtraDR** [comment]
@rpherbig Close?
